### PR TITLE
[feat] StableMemory: load/store implementations for numeric types beyond Nat32

### PR DESCRIPTION
**Pull request by crusso**
- [x] Nat8
- [x] Nat16
- [x] Nat32
- [x] Nat64
- [x] Float
- [x] Int8
- [x] Int16
- [x] Int32
- [x] Int64

### CODE REVIEW

**Comment by nomeata** [comment]
These operations look awfully useful when doing binary data formats. Can we maybe create a “mutable blob” type, with ways to create them backed by stable memory or normal memory, and have a common interface?

(Maybe later of cause)

**Comment by crusso** [comment]
> These operations look awfully useful when doing binary data formats. Can we maybe create a “mutable blob” type, with ways to create them backed by stable memory or normal memory, and have a common interface?
> 
> (Maybe later of cause)

Yeah, I've been wondering about adding a mutable blob type or perhaps we should optimize arrays more generally.

 I'm not sure we want to have each operation dispatch on the type of address, though do we? Also, the load operations would make sense on both mutable and immutable blobs, but I don't think we have any way to abstract on that except overloading (as with our two array flavours).
```
loadNat8 : ({#Blob : Blob; #VarBlob : VarBlob, #Mem}, offset) : Nat8 
storeNat8 : ({#Blob : Blob; #VarBlob : VarBlob, #Mem}, offset, value : Nat8) : Nat8 
```

Looks a bit clunky to me and even less efficient.

Perhaps we should just treat StableMemory as a singleton object and have a class mapping (mutable) Blobs to object with various checked (store/)load  operations.

```
class Reader(b : Blob) {
   loadNat8 : (offset : Nat32) -> Nat8
  ...
}
class ReaderWriter(b : VarBlob) {
 loadNat8 : (offset : Nat32) -> Nat8
 storeNat8 : (offset : Nat32, value: Nat8) -> ()
};
}
```
Which also begs the question of whether to use `Nat32` or `Nat` or maybe just `Nat64` for offsets.

**Comment by crusso** [comment]
Kinda makes you wish you had type classes...

**Review by rossberg — APPROVED** [review]
Cool, LGTM. Maybe @ggreif can have a second look at the codegen.

File: src/codegen/compile.ml [hunk 1]
```diff
@@ -7784,16 +8006,86 @@ and compile_exp (env : E.t) ae exp =
...
+      SR.Vanilla,
+      compile_exp_as env ae SR.UnboxedWord32 e ^^
+      StableMem.load_word8 env ^^
+      TaggedSmallWord.msb_adjust Type.Int8
+
+    | OtherPrim ("stableMemoryStoreNat8"), [e1; e2] ->
+      SR.unit,
+      compile_exp_as env ae SR.UnboxedWord32 e1 ^^
+      compile_exp_as env ae SR.Vanilla e2 ^^ TaggedSmallWord.lsb_adjust Type.Nat8 ^^
+      StableMem.store_word8 env
```
Comment (ggreif) [inline src/codegen/compile.ml:8036]: I wonder if you could sink the `m/lsb_adjusts` into the `StableMem.store_word8`.

File: src/prelude/prim.mo [hunk 2]
```diff
@@ -292,6 +292,54 @@ func stableMemoryGrow(pages : Nat32) : Nat32 =
...
+
+func stableMemoryStoreNat16(offset : Nat32, val : Nat16) : () =
+  (prim "stableMemoryStoreNat16" : (Nat32, Nat16) -> ()) (offset, val);
+
+func stableMemoryLoadNat64(offset : Nat32) : Nat64 =
+  (prim "stableMemoryLoadNat64" : Nat32 -> Nat64) offset;
+
+func stableMemoryStoreNat64(offset : Nat32, val : Nat64) : () =
+  (prim "stableMemoryStoreNat64" : (Nat32, Nat64) -> ()) (offset, val);
+
```
Comment (ggreif) [inline src/prelude/prim.mo:315]: Nit: extra line?

File: src/codegen/compile.ml [hunk 3]
```diff
@@ -3980,6 +4095,113 @@ module StableMem = struct
...
+        (("offset", I32Type)) [F64Type]
+        (fun env get_offset  ->
+          get_offset ^^
+          compile_unboxed_const 8l ^^
+          guard_range env ^^
+          get_offset ^^
+          read_float64 env)
+    | _ -> assert false
+
+  let store_float64 env =
```
Comment (ggreif) [inline src/codegen/compile.ml:4191]: Around line 3936 there is a similar function? What is the semantic difference between `__stablemem_store_float64` and `__stablemem_write_float64`?

File: src/codegen/compile.ml [hunk 4]
```diff
@@ -3832,6 +3832,120 @@ module StableMem = struct
...
+            get_temp_ptr ^^
+            G.i (Load {ty = F64Type; align = 0; offset = 0l; sz = None })))
+    | _ -> assert false
+
+  (* write stable mem offset and value, operands on stack *)
+  let write_float64 env =
+    match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.share_code2 env "__stablemem_write_float64"
+        (("offset", I32Type),("value", F64Type)) []
```
Comment (ggreif) [inline src/codegen/compile.ml:3940]:
```suggestion
        (("offset", I32Type), ("value", F64Type)) []
```
Nit, there are a few more above.

File: src/codegen/compile.ml [hunk 5]
```diff
@@ -3980,6 +4095,113 @@ module StableMem = struct
...
     | _ -> assert false
+
+  let load_word8 env =
+    match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.share_code1 env "__stablemem_load_word8"
+        (("offset", I32Type)) [I32Type]
+        (fun env get_offset  ->
+          get_offset ^^
+          guard env ^^
```
Comment (ggreif) [inline src/codegen/compile.ml:4106]: `guard env` could leave the offset on the stack, then the next line is not necessary.

File: src/codegen/compile.ml (hunk 3 above)
Comment (ggreif) [reply src/codegen/compile.ml:4191]: I presume the `load`/`store` ones operate on guarded ranges?

File: src/codegen/compile.ml [hunk 6]
```diff
@@ -3832,6 +3832,120 @@ module StableMem = struct
...
 
+  (* read word8 from stable mem offset on stack *)
+  let read_word8 env =
+    match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.share_code1 env "__stablemem_read_word8"
+        ("offset", I32Type) [I32Type]
+        (fun env get_offset ->
+          Stack.with_words env "temp_ptr" 1l (fun get_temp_ptr ->
+            get_temp_ptr ^^ get_offset ^^  compile_unboxed_const 1l ^^
```
Comment (ggreif) [inline src/codegen/compile.ml:3843]:
```suggestion
            get_temp_ptr ^^ get_offset ^^ compile_unboxed_const 1l ^^
```
Nit, there are a few around.

**Review by ggreif — APPROVED** [review]
Codegen looks good, a few suggestions and nits only.

File: src/codegen/compile.ml (hunk 1 above)
Comment (ggreif) [reply src/codegen/compile.ml:8036]: Then some of these could be collapsed using `or` patterns.

File: src/prelude/prim.mo (hunk 2 above)
Comment (crusso) [reply src/prelude/prim.mo:315]:
```suggestion
```

File: src/codegen/compile.ml (hunk 1 above)
Comment (crusso) [reply src/codegen/compile.ml:8036]: Yeah, I'm not sure Its worth it . I don't want to retest the string lit and don't want to rely on the type of e2 (that could, in principal, be `Non`). I'll just leave it, but thanks for the suggestion.

File: src/codegen/compile.ml (hunk 5 above)
Comment (crusso) [reply src/codegen/compile.ml:4106]: That's just pushing the load elsewhere and I'd like to keep the code similar to guard_range...

**Pull request merged**